File: fe_book_food/src/types/food.ts
import { CategoryResponse } from "./category";
//phần dùng cho backend
export interface Food{
    id: number;
  name: string;
  price: number;
  image: string;
}

export interface CreateFoodRequest{
    name: string;
    description?: string; //thuôc tính? : tùy chọn( kh bắt buộc)
    price: number;
    imageUrl?: string;
    categoryId: number;
    preparationTime: number;
}

export interface UpdateFoodRequest{
    name: string;
    description?: string; //thuôc tính? : tùy chọn( kh bắt buộc)
    price: number;
    imageUrl?: string;
    categoryId: number;
    preparationTime: number;
    isAvailable?: boolean;
}

export interface FoodSearchParams{
    keyword?: string;
    categoryId? : number;
    minPrice? : number;
    maxPrice?: number;
    page?: number;
    size?: number;
    isAvailable?: boolean;
}

// DTO trả về từ backend
export interface FoodResponse {
  id: number;
  name: string;
  description: string;
  price: number;
  imageUrl: string;
  isAvailable: boolean;
  preparationTime: number;
  rating: number;
  category: CategoryResponse;
  soldCount: number;
  createdAt: string;  // LocalDateTime -> string
  updatedAt: string;  // LocalDateTime -> string
}



export interface FoodRatingSummaryResponse{
    averageRating: number;
    totalReviews: number;
}

File: BE_Book_Food/src/main/java/com/example/ecommerce/book_food/entity/Food.java
package com.example.ecommerce.book_food.entity;

import jakarta.persistence.*;
import jakarta.validation.constraints.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "foods")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Food {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false, length = 100)
    @NotBlank(message = "Food name cannot be blank")
    @Size(max = 100, message = "Food name must not exceed 100 characters")
    private String name;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Column(nullable = false, precision = 10, scale = 2)
    @NotNull(message = "Price cannot be null")
    @Positive(message = "Price must be positive") //đảm bảo giá trị > 0
    private BigDecimal price;

    @Column(name = "image_url")
    private String imageUrl;

    @Column(name = "is_available")
    @NotNull(message = "Availability status cannot be null")
    private Boolean isAvailable = true;

    @Column(name = "preparation_time")
    @NotNull(message = "Preparation time cannot be null")
    @PositiveOrZero(message = "Preparation time must be non-negative")
    private Integer preparationTime = 15; // minutes

    @Column(precision = 2, scale = 1)
    @Builder.Default
    private BigDecimal rating = BigDecimal.ZERO;

    @Column(name = "sold_count", nullable = false)
    private long soldCount = 0L;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "category_id")
    private Category category;

    @OneToMany(mappedBy = "food", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    @Builder.Default
    private List<OrderItem> orderItems = new ArrayList<>();

    @OneToMany(mappedBy = "food", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    @Builder.Default
    private List<Review> reviews = new ArrayList<>();

    @CreationTimestamp
    @Column(name = "created_at", updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;


}